creusot-contracts/src/logic/ord.rs

```diff
@@ -0,0 +1,66 @@
+use crate as creusot_contracts;
+use creusot_contracts_proc::*;
+
+use crate::logic::EqLogic;
+pub use std::cmp::Ordering;
+
+#[logic]
+pub fn le_log<T: OrdLogic + ?Sized>(a: T, o: T) -> bool {
+    match a.cmp_log(o) {
+        Ordering::Greater => false,
+        _ => true,
+    }
+}
+
+#[logic]
+pub fn ge_log<T: OrdLogic + ?Sized>(a: T, o: T) -> bool {
+    match a.cmp_log(o) {
+        Ordering::Less => false,
+        _ => true,
+    }
+}
+#[logic]
+pub fn gt_log<T: OrdLogic + ?Sized>(a: T, o: T) -> bool {
```

Comment:
the big disadvantage of moving this to a top-level function is that it's no longer possible to override it for integer types so that we emit `<` instead.

Reply:
(this is a **very** minor issue though.

Reply:
I wish why3 had the equivalent of Coq's notations so at least we could pretty-print in proof tasks.

Reply:
That's indeed a very important point, since Why3's `<` is translated to the provers' native comparison operator over integers, and I guess this is crucial for good performance of provers. So I guess this forces us to use default implementations, and require a lemma that `<` is indeed equivalent to `cmp`.

BTW, we should be very cautious about default specifications of default implementations, as it can cause unsoundness. Indeed, if the specification of a default implementation of a lemma/program function depends on a logic function/predicate which has a default implementation, then we should ask the user to reprove the specification of the lemma/program funciton as soon as either the lemma/program function OR logic function/predicate is given a non-default implementation.

In this particular case, let's call `cmp_gt_log` the lemma stating that `cmp` and `gt_log` are equivalent. This means that we should make sure that we generate VCs for `cmp_gt_log` as soon as we give a non-default implementation for `cmp_gt_log` (of course) OR `gt_log` (which is less obvious).

Reply:
Agreed, if we provide default logical methods we need to ensure coherence between definitions.

Reply:
Fwiw, I think we should *always* generate vcs, but if you don't change any of the definitions involved the proof should be trivial (just what was proven for the trait).

Reply:
That's a good first implementation, yes.

But imagine a default implementation of program function. Its body can be quite complex, so a large number of VCs could be generated from that, and perhaps the proof is not even automatic. If we don't change anything, then we'd like to avoid re-doing the proof (that's the whole point of default implementations !).

The problem here is that we should make sure the spec does not depend on a reimplemented default implementation, transitively (if it depends on a default implementation which depends on a method which had a defaulf implementation which has been overridden, then the proof is out-of-date). So I guess this criterion is rather subtle to implement.

Reply:
If we don't change anything we literally clone the default substitution so we inherit the proof as well!

Reply:
Ah i see your concern, yes we could sneak in unsoundness during this clone process if another default implementation was also changed.

Reply:
> The problem here is that we should make sure the spec does not depend on a reimplemented default implementation, transitively (if it depends on a default implementation which depends on a method which had a defaulf implementation which has been overridden, then the proof is out-of-date). So I guess this criterion is rather subtle to implement.

I believe we can implement this by asking if the proof (transitively) depends on the implementation. If it does we somehow force a new proof.